etc/migration/migrateUsersToShibboleth.js

```diff
@@ -0,0 +1,352 @@
+    _userHashes = _userHashes || [];
+
+    if (_.isEmpty(userHashes)) {
+        return callback(_userHashes);
+    }
+
+    // Take the next principal hash from the collection
+    var principalHash = userHashes.shift();
+    _findExternalId(principalHash, function(err, userHash) {
+        if (!err && userHash) {
+            // Accumulate the return information we want
+            _userHashes.push(userHash);
+        }
+
+        // Our recursive step inside the callback for _findExternalId
+        return _getExternalIds(userHashes, callback, _userHashes);
+    });
+};
+
+/**
+ * Find the external login id for the user and add it to the user hash
+ *
+ * @param  {Object}     userHash                The user hash object
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {String}     callback.userHash       The user hash object including the external id
+ * @api private
+ */
+function _findExternalId(userHash, callback) {
+    var userId = userHash.principalId;
+    Cassandra.runQuery('SELECT "loginId" FROM "AuthenticationUserLoginId" WHERE "userId" = ?', [userId], function(err, rows) {
+        if (err) {
+            log().error({'err': err}, 'Failed Cassandra query for user %s\'s loginId', userHash.email);
+            _writeErrorRow(user, 'Could not retrieve loginId for this user');
+            return callback(err);
+        }
+
+        var loginId = _.chain(rows)
+            .map(Cassandra.rowToHash)
+            .pluck('loginId')
+            .first()
+            .value();
+
+        return callback(null, _.extend(userHash, {'loginId': loginId}));
+    });
+}
+
+/**
+ * Create new authentication records linked to Shibboleth EPPN for all users within the tenant
+ *
+ * @param  {Object[]}   allUsers                The array of all the users for a tenancy
+ * @param  {Function}   callback                Standard callback function
+ * @param  {Object}     callback.err            An error that occurred, if any
+ * @param  {Object[]}   callback.mappedUsers    The Shibboleth users that were mapped to existing users
+ * @api private
+ */
+var _mapUsersToShibLogin = function(allUsers, callback) {
+    // Remove users with duplicate emails or invalid login IDs
+    var cleanedUsers = _.chain(allUsers)
+        .filter(function(user) {
+            var valid = user.loginId;
+            if (!valid) {
+                _writeErrorRow(user, 'This user has an invalid loginId');
+            }
+            return valid;
+        })
+        .groupBy(function(user) {
+            return user.email;
+        })
+        .map(function(grouped) {
+            if (grouped.length === 1) {
+                return grouped[0];
+            } else {
+                _.each(grouped, function(dupe) {
+                    _writeErrorRow(dupe, 'Found more than one user with this email');
+                });
+            }
+        })
```

Comment:
Because we don't know which email we should map? Plus I assumed that if we added two identical `loginId`s into the AuthenticationLoginId table, Cassandra would be unhappy.